Add neg bit

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59380 Add fast path for addmm when the inputs are conjugate
* **#56058 Add neg bit**

User facing changes:
1. Adds a negative bit and corresponding new API (`is_neg()`,`resolve_neg()`)
2. `tensor.conj().imag` now returns a floating point tensor with neg bit set to 1 instead of a tensor with no notion of negative bit. Note that imag is still a view and all the view properties still hold for imag.

Non user facing changes:
1. Added a new Negative dispatch key and a backend fallback to handle it
2. Updated copy kernel to handle negative bit
3. Merged conjugate and negative bit fallback kernel
4. fixed https://github.com/pytorch/pytorch/issues/60478 (caused due to https://github.com/pytorch/pytorch/pull/54987)

Testing:
1. Added a new OpInfo based test `test_neg_view` (verifies that out-of-place and in-place operations work correctly for all operations when the input is a neg view tensor by checking the result against an actually negated tensor, verifies that autograd returns the same output for both neg view and actually negated tensors as well as it works fine when grad_out is a neg view).
2. Added a new test class containing `test_conj_view`, `test_neg_view`. 

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D29636403](https://our.internmc.facebook.com/intern/diff/D29636403)

anjali411 committed:
Make negative operation a view

[ghstack-poisoned]

anjali411 committed:
Update on "Make negative operation a view"




[ghstack-poisoned]

anjali411 committed:
Update on "Make negative operation a view"




[ghstack-poisoned]

anjali411 committed:
Update on "Make negative operation a view"




[ghstack-poisoned]

anjali411 committed:
Update on "Make negative operation a view"




[ghstack-poisoned]

anjali411 committed:
Update on "Make negative operation a view"




[ghstack-poisoned]

anjali411 committed:
Update on "Make negative operation a view"




[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/NegateFallback.cpp:
```diff
@@ -0,0 +1,122 @@
+    native::neg_(mutable_input);
+    mutable_input.set_neg(true);
+  }
+}
```
As the logic here is substantively the same, it would be nice if we could reuse the implementation in the two cases. Maybe you can template over the functions you need to call (e.g., `resolve_neg` and `set_neg` and `neg_`) and then instantiate it twice. Maybe a class you can inherit from and then fill in the meaning of some functions would work OK too.

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "




[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


TODO: 
1. remove view_as_real_physical
2. reuse ConjugateFallback code


[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


TODO: 
1. remove view_as_real_physical
2. reuse ConjugateFallback code


[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


TODO: 
1. remove view_as_real_physical
2. reuse ConjugateFallback code


[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


TODO: 
1. remove view_as_real_physical
2. reuse ConjugateFallback code


[ghstack-poisoned]

anjali411 commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -344,10 +344,23 @@ Tensor real(const Tensor& self) {
   }
 }
 
+Tensor neg_view(const Tensor& self) {
```
@ezyang should we make this API public to let users get negative view if they want (even when not in complex world)?

ezyang replied:
Maybe with a leading underscore. I wouldn't really advertise it all that much as there probably isn't very much benefit at all you can get from it.

ezyang commented on aten/src/ATen/native/MathBitsFallback.cpp:
```diff
@@ -0,0 +1,119 @@
+
+namespace at {
+
+struct MathOpFallback {
```
I wonder if "MathOpFallback" is really the right way to talk about this. I think the more important property you need for the algorithm here to be correct is for the op that we're doing the view on to be "self inverse", since we use the self inverse property to mean that we only need a single boolean to represent the state (false -f-> true -f-> false).

ezyang replied:
Yes. But I don't think this class works with non-self-inverse but invertible ops. In particular, you need both `math_op_` and `math_op_inverse_` in that case. Nor do I think it makes sense to prematurely generalize this class to support it.

ezyang commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -1184,7 +1184,7 @@ TORCH_LIBRARY_IMPL(aten, Batched, m) {
   TRIVIAL_OP(_conj);
   TRIVIAL_OP(resolve_conj);
-  TRIVIAL_OP(_resolve_conj);
+  TRIVIAL_OP(_resolve_conj_neg);
```
Although it "doesn't really matter" because this is a private underscored function, it is weird naming, especially if you imagine us ever adding other view bits the name would keep getting longer and longer.

ezyang replied:
This sort of relates to a discussion we had earlier about whether or not new/like should propagate conjugate/negative bits. Anything that *doesn't* get propagated gets "cleared" by the resolve method implemented by allocating a new tensor and copy_'ing into it. Is this the semantics of `clone()`? Here is clone:

```
Tensor clone(const Tensor& src, c10::optional<c10::MemoryFormat> optional_memory_format) {
  auto memory_format =
      optional_memory_format.value_or(MemoryFormat::Preserve);
  if (memory_format == MemoryFormat::Preserve) {
    if (src.is_non_overlapping_and_dense()) {
      // Copy all strides
      auto self = at::empty_strided(src.sizes(), src.strides(), src.options());
      self.copy_(src);
      return self;
    } else {
      memory_format = src.suggest_memory_format();
    }
  }
  auto self = at::empty_like(src, src.options(), memory_format);
  self.copy_(src);
  return self;
}
```

so really maybe just deferring to `clone` is sufficient.

anjali411 replied:
yeah good point! we can just use clone here

ezyang commented on aten/src/ATen/native/MathBitsFallback.cpp:
```diff
@@ -0,0 +1,119 @@
+
+struct MathOpFallback {
+  MathOpFallback(DispatchKey key_, string op_name_) : key(key_), op_name(op_name_) {}
+  virtual bool is_bit_set(c10::TensorImpl&) = 0;
```
There isn't really any reason for this to take a mutable reference. Also the TensorImpl here is weirdly asymmetric with the Tensor everywhere else; maybe there is a good reason for this but it is not immediately evident.

ezyang commented on aten/src/ATen/native/MathBitsFallback.cpp:
```diff
@@ -0,0 +1,119 @@
+struct MathOpFallback {
+  MathOpFallback(DispatchKey key_, string op_name_) : key(key_), op_name(op_name_) {}
+  virtual bool is_bit_set(c10::TensorImpl&) = 0;
+  virtual void set_bit(Tensor&, bool) = 0;
```
These (all the `Tensor&` references) also don't have to be mutable references, but for a different reason: const-correctness on Tensor is a lie and `const Tensor&` will always have the same set of methods available.

ezyang replied:
https://github.com/zdevito/ATen/issues/27#issuecomment-330717839

ezyang commented on aten/src/ATen/native/MathBitsFallback.cpp:
```diff
@@ -0,0 +1,119 @@
+
+namespace at {
+
+struct MathOpFallback {
```
Strongly recommend leaving a comment at the top of the class giving a general idea about what this abstracts over and how to use the abstraction.

anjali411 replied:
done!

ezyang commented on aten/src/ATen/native/MathBitsFallback.cpp:
```diff
@@ -0,0 +1,119 @@
+      }
+      if (ivalue.isTensor()) {
+        auto* impl = ivalue.unsafeToTensorImpl();
+        if (!is_bit_set(*impl)) {
```
yeah, thanks to @swolchok, you can get a `const Tensor&` from the ivalue at no cost, so might as well do that.

ezyang replied:
(this wasn't true when this code was originally written)

anjali411 replied:
makes sense!

ezyang commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -344,10 +344,23 @@ Tensor real(const Tensor& self) {
-    return at::select(real_tensor, real_tensor.dim() - 1, 1);
+    auto real_tensor = at::_view_as_real_physical(self);
+    auto true_real_tensor = self.is_conj() ? at::neg_view(real_tensor) : real_tensor;
+    return at::select(true_real_tensor, real_tensor.dim() - 1, 1);
```
nit: this is a preexisting condition, but it ought to be possible to skip redundant redispatches here

anjali411 replied:
there doesn't seem to be a dispatch for `select` (forward).

ezyang commented on aten/src/ATen/native/cpu/CopyKernel.cpp:
```diff
@@ -23,41 +23,71 @@ static void copy_kernel(TensorIterator& iter, bool non_blocking) {
+                [=](Vec256<scalar_t> a) -> Vec256<scalar_t> { return a.neg(); });
+            });
+        } else {
+          AT_DISPATCH_COMPLEX_TYPES(dtype, "conj_kernel", [&] {
```
It's worth saying that this case of the condition should ~never actually happen (unless someone does `neg_view` but this is not a real thing)

anjali411 committed:
Update on "Add neg bit "


TODO: 
1. remove view_as_real_physical
2. reuse ConjugateFallback code


[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


TODO: 
1. remove view_as_real_physical
2. reuse ConjugateFallback code


[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


TODO: 
1. remove view_as_real_physical
2. reuse ConjugateFallback code


[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


TODO: 
1. remove view_as_real_physical
2. reuse ConjugateFallback code


[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


TODO: 
1. remove view_as_real_physical
2. reuse ConjugateFallback code


[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


TODO: 
1. remove view_as_real_physical
2. reuse ConjugateFallback code


[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


TODO: 
1. remove view_as_real_physical
2. reuse ConjugateFallback code


[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/).

anjali411 committed:
Update on "Add neg bit "


TODO: 
1. remove view_as_real_physical
2. reuse ConjugateFallback code

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


TODO: 
1. remove view_as_real_physical
2. reuse ConjugateFallback code

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/cuda/Copy.cu:
```diff
@@ -72,16 +73,30 @@ void copy_device_to_device(TensorIterator& iter, bool non_blocking) {
+          AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(
+              kHalf, kBool, kBFloat16, dtype, "copy_", [&] {
+                gpu_kernel(iter, [] GPU_LAMBDA(scalar_t x) { return -x; });
+              });
```
At some point we gonna have to have a chat about our O(2^n) conversions LOL

anjali411 replied:
yeah

anjali411 committed:
Update on "Add neg bit "


TODO: 
1. remove view_as_real_physical
2. reuse ConjugateFallback code

bc breaking: sparse_tensor.conj().imag doesn't work anymore.
Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


User facing changes:
1. Adds a negative bit and corresponding new API (`is_neg()`,`resolve_neg()`)
2. `tensor.conj().imag` now returns a floating point tensor with neg bit set to 1 instead of a tensor with no notion of negative bit. Note that imag is still a view and all the view properties still hold for imag.

Non user facing changes:
1. Added a new Negative dispatch key and a backend fallback to handle it
2. Updated copy kernel to handle negative bit
3. Merged conjugate and negative bit fallback kernel
4. fixed https://github.com/pytorch/pytorch/issues/60478 (caused due to https://github.com/pytorch/pytorch/pull/54987)

Testing:
1. Added a new OpInfo based test `test_neg_view` (verifies that out-of-place and in-place operations work correctly for all operations when the input is a neg view tensor by checking the result against an actually negated tensor, verifies that autograd returns the same output for both neg view and actually negated tensors as well as it works fine when grad_out is a neg view).
2. Added a new test class containing `test_conj_view`, `test_neg_view`. 

TODO:
follow up on https://github.com/pytorch/pytorch/pull/60522#issuecomment-869353294

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


User facing changes:
1. Adds a negative bit and corresponding new API (`is_neg()`,`resolve_neg()`)
2. `tensor.conj().imag` now returns a floating point tensor with neg bit set to 1 instead of a tensor with no notion of negative bit. Note that imag is still a view and all the view properties still hold for imag.

Non user facing changes:
1. Added a new Negative dispatch key and a backend fallback to handle it
2. Updated copy kernel to handle negative bit
3. Merged conjugate and negative bit fallback kernel
4. fixed https://github.com/pytorch/pytorch/issues/60478 (caused due to https://github.com/pytorch/pytorch/pull/54987)

Testing:
1. Added a new OpInfo based test `test_neg_view` (verifies that out-of-place and in-place operations work correctly for all operations when the input is a neg view tensor by checking the result against an actually negated tensor, verifies that autograd returns the same output for both neg view and actually negated tensors as well as it works fine when grad_out is a neg view).
2. Added a new test class containing `test_conj_view`, `test_neg_view`. 

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


User facing changes:
1. Adds a negative bit and corresponding new API (`is_neg()`,`resolve_neg()`)
2. `tensor.conj().imag` now returns a floating point tensor with neg bit set to 1 instead of a tensor with no notion of negative bit. Note that imag is still a view and all the view properties still hold for imag.

Non user facing changes:
1. Added a new Negative dispatch key and a backend fallback to handle it
2. Updated copy kernel to handle negative bit
3. Merged conjugate and negative bit fallback kernel
4. fixed https://github.com/pytorch/pytorch/issues/60478 (caused due to https://github.com/pytorch/pytorch/pull/54987)

Testing:
1. Added a new OpInfo based test `test_neg_view` (verifies that out-of-place and in-place operations work correctly for all operations when the input is a neg view tensor by checking the result against an actually negated tensor, verifies that autograd returns the same output for both neg view and actually negated tensors as well as it works fine when grad_out is a neg view).
2. Added a new test class containing `test_conj_view`, `test_neg_view`. 

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


User facing changes:
1. Adds a negative bit and corresponding new API (`is_neg()`,`resolve_neg()`)
2. `tensor.conj().imag` now returns a floating point tensor with neg bit set to 1 instead of a tensor with no notion of negative bit. Note that imag is still a view and all the view properties still hold for imag.

Non user facing changes:
1. Added a new Negative dispatch key and a backend fallback to handle it
2. Updated copy kernel to handle negative bit
3. Merged conjugate and negative bit fallback kernel
4. fixed https://github.com/pytorch/pytorch/issues/60478 (caused due to https://github.com/pytorch/pytorch/pull/54987)

Testing:
1. Added a new OpInfo based test `test_neg_view` (verifies that out-of-place and in-place operations work correctly for all operations when the input is a neg view tensor by checking the result against an actually negated tensor, verifies that autograd returns the same output for both neg view and actually negated tensors as well as it works fine when grad_out is a neg view).
2. Added a new test class containing `test_conj_view`, `test_neg_view`. 

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29636403).

anjali411 committed:
Update on "Add neg bit "


User facing changes:
1. Adds a negative bit and corresponding new API (`is_neg()`,`resolve_neg()`)
2. `tensor.conj().imag` now returns a floating point tensor with neg bit set to 1 instead of a tensor with no notion of negative bit. Note that imag is still a view and all the view properties still hold for imag.

Non user facing changes:
1. Added a new Negative dispatch key and a backend fallback to handle it
2. Updated copy kernel to handle negative bit
3. Merged conjugate and negative bit fallback kernel
4. fixed https://github.com/pytorch/pytorch/issues/60478 (caused due to https://github.com/pytorch/pytorch/pull/54987)

Testing:
1. Added a new OpInfo based test `test_neg_view` (verifies that out-of-place and in-place operations work correctly for all operations when the input is a neg view tensor by checking the result against an actually negated tensor, verifies that autograd returns the same output for both neg view and actually negated tensors as well as it works fine when grad_out is a neg view).
2. Added a new test class containing `test_conj_view`, `test_neg_view`. 

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D29636403](https://our.internmc.facebook.com/intern/diff/D29636403)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29636403).

anjali411 committed:
Update on "Add neg bit "


User facing changes:
1. Adds a negative bit and corresponding new API (`is_neg()`,`resolve_neg()`)
2. `tensor.conj().imag` now returns a floating point tensor with neg bit set to 1 instead of a tensor with no notion of negative bit. Note that imag is still a view and all the view properties still hold for imag.

Non user facing changes:
1. Added a new Negative dispatch key and a backend fallback to handle it
2. Updated copy kernel to handle negative bit
3. Merged conjugate and negative bit fallback kernel
4. fixed https://github.com/pytorch/pytorch/issues/60478 (caused due to https://github.com/pytorch/pytorch/pull/54987)

Testing:
1. Added a new OpInfo based test `test_neg_view` (verifies that out-of-place and in-place operations work correctly for all operations when the input is a neg view tensor by checking the result against an actually negated tensor, verifies that autograd returns the same output for both neg view and actually negated tensors as well as it works fine when grad_out is a neg view).
2. Added a new test class containing `test_conj_view`, `test_neg_view`. 

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D29636403](https://our.internmc.facebook.com/intern/diff/D29636403)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29636403).

anjali411 committed:
Update on "Add neg bit "


User facing changes:
1. Adds a negative bit and corresponding new API (`is_neg()`,`resolve_neg()`)
2. `tensor.conj().imag` now returns a floating point tensor with neg bit set to 1 instead of a tensor with no notion of negative bit. Note that imag is still a view and all the view properties still hold for imag.

Non user facing changes:
1. Added a new Negative dispatch key and a backend fallback to handle it
2. Updated copy kernel to handle negative bit
3. Merged conjugate and negative bit fallback kernel
4. fixed https://github.com/pytorch/pytorch/issues/60478 (caused due to https://github.com/pytorch/pytorch/pull/54987)

Testing:
1. Added a new OpInfo based test `test_neg_view` (verifies that out-of-place and in-place operations work correctly for all operations when the input is a neg view tensor by checking the result against an actually negated tensor, verifies that autograd returns the same output for both neg view and actually negated tensors as well as it works fine when grad_out is a neg view).
2. Added a new test class containing `test_conj_view`, `test_neg_view`. 

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D29636403](https://our.internmc.facebook.com/intern/diff/D29636403)

[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


User facing changes:
1. Adds a negative bit and corresponding new API (`is_neg()`,`resolve_neg()`)
2. `tensor.conj().imag` now returns a floating point tensor with neg bit set to 1 instead of a tensor with no notion of negative bit. Note that imag is still a view and all the view properties still hold for imag.

Non user facing changes:
1. Added a new Negative dispatch key and a backend fallback to handle it
2. Updated copy kernel to handle negative bit
3. Merged conjugate and negative bit fallback kernel
4. fixed https://github.com/pytorch/pytorch/issues/60478 (caused due to https://github.com/pytorch/pytorch/pull/54987)

Testing:
1. Added a new OpInfo based test `test_neg_view` (verifies that out-of-place and in-place operations work correctly for all operations when the input is a neg view tensor by checking the result against an actually negated tensor, verifies that autograd returns the same output for both neg view and actually negated tensors as well as it works fine when grad_out is a neg view).
2. Added a new test class containing `test_conj_view`, `test_neg_view`. 

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D29636403](https://our.internmc.facebook.com/intern/diff/D29636403)

[ghstack-poisoned]

anjali411 committed:
Update on "Add neg bit "


User facing changes:
1. Adds a negative bit and corresponding new API (`is_neg()`,`resolve_neg()`)
2. `tensor.conj().imag` now returns a floating point tensor with neg bit set to 1 instead of a tensor with no notion of negative bit. Note that imag is still a view and all the view properties still hold for imag.

Non user facing changes:
1. Added a new Negative dispatch key and a backend fallback to handle it
2. Updated copy kernel to handle negative bit
3. Merged conjugate and negative bit fallback kernel
4. fixed https://github.com/pytorch/pytorch/issues/60478 (caused due to https://github.com/pytorch/pytorch/pull/54987)

Testing:
1. Added a new OpInfo based test `test_neg_view` (verifies that out-of-place and in-place operations work correctly for all operations when the input is a neg view tensor by checking the result against an actually negated tensor, verifies that autograd returns the same output for both neg view and actually negated tensors as well as it works fine when grad_out is a neg view).
2. Added a new test class containing `test_conj_view`, `test_neg_view`. 

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D29636403](https://our.internmc.facebook.com/intern/diff/D29636403)

[ghstack-poisoned]

ezyang commented on torch/_tensor_str.py:
```diff
@@ -233,10 +233,15 @@ def _tensor_str(self, indent):
     summarize = self.numel() > PRINT_OPTS.threshold
+
+    # handle the negative bit
+    self = self.resolve_neg()
```
Maybe it's this.

anjali411 committed:
Update on "Add neg bit "


User facing changes:
1. Adds a negative bit and corresponding new API (`is_neg()`,`resolve_neg()`)
2. `tensor.conj().imag` now returns a floating point tensor with neg bit set to 1 instead of a tensor with no notion of negative bit. Note that imag is still a view and all the view properties still hold for imag.

Non user facing changes:
1. Added a new Negative dispatch key and a backend fallback to handle it
2. Updated copy kernel to handle negative bit
3. Merged conjugate and negative bit fallback kernel
4. fixed https://github.com/pytorch/pytorch/issues/60478 (caused due to https://github.com/pytorch/pytorch/pull/54987)

Testing:
1. Added a new OpInfo based test `test_neg_view` (verifies that out-of-place and in-place operations work correctly for all operations when the input is a neg view tensor by checking the result against an actually negated tensor, verifies that autograd returns the same output for both neg view and actually negated tensors as well as it works fine when grad_out is a neg view).
2. Added a new test class containing `test_conj_view`, `test_neg_view`. 

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D29636403](https://our.internmc.facebook.com/intern/diff/D29636403)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29636403).

anjali411 committed:
Update on "Add neg bit "


User facing changes:
1. Adds a negative bit and corresponding new API (`is_neg()`,`resolve_neg()`)
2. `tensor.conj().imag` now returns a floating point tensor with neg bit set to 1 instead of a tensor with no notion of negative bit. Note that imag is still a view and all the view properties still hold for imag.

Non user facing changes:
1. Added a new Negative dispatch key and a backend fallback to handle it
2. Updated copy kernel to handle negative bit
3. Merged conjugate and negative bit fallback kernel
4. fixed https://github.com/pytorch/pytorch/issues/60478 (caused due to https://github.com/pytorch/pytorch/pull/54987)

Testing:
1. Added a new OpInfo based test `test_neg_view` (verifies that out-of-place and in-place operations work correctly for all operations when the input is a neg view tensor by checking the result against an actually negated tensor, verifies that autograd returns the same output for both neg view and actually negated tensors as well as it works fine when grad_out is a neg view).
2. Added a new test class containing `test_conj_view`, `test_neg_view`. 

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D29636403](https://our.internmc.facebook.com/intern/diff/D29636403)

[ghstack-poisoned]

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29636403).
